### DISCUSSION TITLE
Run dagger (python) with a local docker image

### DISCUSSION THREAD

**Discussion by MathKlim** [dagger/dagger#5147] [Help]
If I use the code given in the doc : 

```py
"""Execute a command."""

import sys
import anyio
import dagger


async def test():
    async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
        python = (
            client.container()
            # pull container
            .from_("ubuntu:22.04")
            # get Python version
            .with_exec(["whoami"])
        )

        # execute
        user = await python.stdout()

    print(f"Hello from Dagger with {user}")


if __name__ == "__main__":
    anyio.run(test)
```

And I run it, I get the answer I expected.

```
Connected to engine 4f1264a89eef
#1 resolve image config for docker.io/library/ubuntu:22.04
#1 DONE 5.3s

#2 from ubuntu:22.04
#2 resolve docker.io/library/ubuntu:22.04@sha256:dfd64a3b4296d8c9b62aa3309984f8620b98d87e47492599ee20739e8eb54fbf done
#2 CACHED

#3 
#3 0.115 root
#3 DONE 0.2s
Hello from Dagger with root
```

But now if I create a Dockerfile and build it :

```dockerfile
FROM ubuntu:22.04

ENTRYPOINT [ "/bin/bash" ]
```

```bash
docker build -t vorphus/test -f Dockerfile .
```

Then the modified version 

```py
"""Execute a command."""

import sys
import anyio
import dagger


async def test():
    async with dagger.Connection(dagger.Config(log_output=sys.stderr)) as client:
        python = (
            client.container()
            # pull container
            .from_("vorphus/test:latest")
            # get Python version
            .with_exec(["whoami"])
        )

        # execute
        user = await python.stdout()

    print(f"Hello from Dagger with {user}")


if __name__ == "__main__":
    anyio.run(test)
```

is unable to retrieve the image, which is built and present locally.

Do I always have to push my containers in a registry to use dagger ? How can I use dagger with local images ?

**Comment by gerhard** [comment]
Does this answer your question? https://docs.dagger.io/252029/load-images-local-docker-engine

**Reply by helderco** [reply]
I think OP wants the inverse.

**Reply by MathKlim** [reply]
> I think OP wants the inverse.

Indeed, that the inverse am I looking for.

**Comment by helderco** [comment]
The simplest is to build the container in Dagger. See [Build image from Dockerfile](https://docs.dagger.io/7442989/cookbook#build-image-from-dockerfile).

If you really need to build outside of Dagger, you can push to a registry first (even if on localhost) so that Dagger can pull from it.

Other than that, I'm not sure if we have a way to import from an OCI `.tar` like we do with export.

**Reply by helderco** [reply]
Within dagger you should also be able to spin up a registry service, use a container with the docker cli to push the image to that registry, and pull from the registry. The registry will be shutdown when no longer needed. This makes it self-contained in dagger, without requiring other tools or processes, and can be abstracted in a function.

**Reply by MathKlim** [reply]
> Within dagger you should also be able to spin up a registry service, use a container with the docker cli to push the image to that registry, and pull from the registry.

Interesting, Do you have any resource I could look for that ? I suppose you're talking about something like that https://docs.dagger.io/7442989/cookbook/#publish-image-to-registry ?

**Reply by helderco** [reply]
If you have an external registry that you want to use, then yes.

But I was referring something more like creating a docker registry service within dagger, then publishing to that: https://docs.dagger.io/757394/use-service-containers#bind-services

I think I've seen that kind of thing around but not sure where. Can't find it in our docs or examples.

May I ask your use case for needing to pull the image from local docker engine to dagger?

**Reply by MathKlim** [reply]
I'm an MLOps/MLE, we (almost) always have several steps in an ML project: preprocess, training, prediction, postprocess. I've designed them so that each one can be built on a separate container, so that if one fail I can just spin up again this one and not the whole flow, so that's fairly classic up to now.

Before pushing them into a container registry, I wanted to test the end-to-end flow via dagger to check if the container are well configured. I can do this by writing some bash script to glue the containers, but it is always a bit clunky with conditionnals between training mode and prediction mode.

My goal was also to permit data scientists who might not have permision to access to the container registry, but who have obviously access to the repo with the Dockerfiles, to use dagger instead of bash to write pipelines and test them locally with local built images.

In the end, my goal is to have a quickier feedback loop rather than passing to some CI pipelines to build, test, register the containers and then run dagger to get the results.

**Reply by helderco** [reply]
Sorry, it's still not clear. Can you use dagger to build the dockerfile, spin up a registry (local to dagger), push the image there, do your tests and if all succeeds then push to the shared container registry?

Or is it not possible to use dagger up to the point where an image is built in the local docker engine?

Or maybe all you want is use to use `docker` cli within dagger which is also possible.

**Reply by MathKlim** [reply]
Hi ! Sorry for the late answer.

* **Can you use dagger to build the dockerfile**:  well, yes, I can do that.

* **spin up a registry (local to dagger), push the image there**: I could do that, but that would include extra steps.  Ultimately, I think my question is the following one. *What If I don't want to bother to create a registry that will be created, used, and destroyed right after ? Why can't I save my image in the same place that the other ones, i.e.  /var/lib/docker/storage_driver ?* From the responses below I understand that dagger isn't built on docker, but another oci runner, so that must be the reason. If that is not an intended use case of dagger that's ok, but that wasn't really clear at all to me reading the documentation.

* **Or maybe all you want is use to use `docker` cli within dagger which is also possible.**: If I can use some plain code and do not depend on some cli (docker, podman, buildah whatever), I'm all for it. That i why dagger looked interesting.

**Comment by helderco** [comment]
@MathKlim sorry, forgot about this. You can import an image from the local docker engine using [Container.import_](https://dagger-io.readthedocs.io/en/sdk-python-v0.6.0/api.html#dagger.api.gen.Container.import_):

Here's an example:
```python
import anyio

import dagger


async def main():
    async with dagger.Connection() as client:
        user = await (
            client.container()
            .with_(docker_image(
                "vorphus/test:latest",
                client.host().unix_socket("/var/run/docker.sock"),
            ))
            .with_exec(["whoami"])
            .stdout()
        )

    print(f"Hello from Dagger with {user}")


def docker_image(name: str, socket: dagger.Socket):
    """Import a docker image into container."""

    def _docker_image(ctr: dagger.Container):
        return ctr.import_(
            ctr.with_(
                docker_exec(
                    ["image", "save", "-o", "image.tar", name],
                    socket,
                )
            )
            .file("image.tar")
        )
    return _docker_image


def docker_exec(args: list[str], socket: dagger.Socket):
    """Execute a docker command."""

    def _docker_exec(ctr: dagger.Container):
        return (
            ctr.from_("docker:23.0.1-cli")
            .with_unix_socket("/var/run/docker.sock", socket)
            .with_exec(args)
        )
    return _docker_exec


if __name__ == "__main__":
    anyio.run(main)
```

Notice that I split reusable functionality into separate functions.

I've added an issue for a guide:
- https://github.com/dagger/dagger/issues/5235

**Reply by dmarteau** [reply]
Hi,

I bumped into the same issue, since our testing CI stages include having local images built by some outer processes.

I cannot figure what prevent dagger to use the local image as the `FROM` directive from any Dockerfile since it based on the same build engine.

Imho, using an tarball image is more or less an totally inefficient anti-pattern since it require to serialize/deserialize data from and to the same engine.

**Reply by helderco** [reply]
@dmarteau Yes, as I mentioned before, the easiest way is to use some registry as intermediary. While dagger shares the same engine as docker build, they are bootstrapped quite differently. When you build with `docker build` it's bootstrapped to work with the docker engine. But dagger isn't dependent on docker specifically, it depends on a container runtime like containerd.

I actually didn't know you could pull in `FROM` from the local engine. I'm curious if it works when using `buildtctl` directly.

\cc @vito @sipsma who can give a better explanation of whether there's a way today to make it work, an idea of how it could work or just an explanation if it's complicated.

**Comment by dmarteau** [comment]
> But dagger isn't dependent on docker specifically

Oh ok, I thought that Dagger was relying on the local docker engine for  building.

> I actually didn't know you could pull in FROM from the local engine

`docker build`  always resolve image (tag or manifest) from the local engine then pull if the image is not found locally - It will be barely usable if we had to rely on registry each time we want to build a set of inherited images.
